m4a_hq_mixer.s

```diff
@@ -717,101 +711,41 @@ C_fast_mixing_creation_loop:
-    ADDS    R5, R5, #0x40000000         @ do that for 4 blocks
+    ADDS    R5, R5, #0x40000000         @ do that for 4 blocks (unused pointer bits)
     BCC     C_fast_mixing_creation_loop
 
 C_skip_fast_mixing_creation:
     LDR     R8, [SP]                        @ restore R8 with the frame length
     LDR     R8, [R8, #(ARG_FRAME_LENGTH + 0x8 + 0xC)]
-    MOV     R2, #0xFF000000                 @ load the fine position overflow bitmask
+    MOVS    R2, #0xFF000000                 @ load the fine position overflow bitmask, set NE
     LDRSB   R12, [R3]
     SUB     R12, R12, R6
 C_fast_mixing_loop:
     /* This is the actual processing and interpolation code loop; NOPs will be replaced by the code above */
-    LDMIA   R5, {R0, R1, R10, LR}       @ load 4 stereo samples to Registers
-    MUL     R9, R7, R12
 fast_mixing_instructions:
-    NOP                                 @ Block #1
-    NOP
-    MLANE   R0, R11, R9, R0
-    NOP
-    NOP
-    NOP
-    NOP
-    BIC     R7, R7, R2, ASR#1
-    MULNE   R9, R7, R12
-    NOP                                 @ Block #2
-    NOP
-    MLANE   R1, R11, R9, R1
-    NOP
-    NOP
-    NOP
-    NOP
-    BIC     R7, R7, R2, ASR#1
-    MULNE   R9, R7, R12
-    NOP                                 @ Block #3
-    NOP
-    MLANE   R10, R11, R9, R10
-    NOP
-    NOP
-    NOP
-    NOP
-    BIC     R7, R7, R2, ASR#1
-    MULNE   R9, R7, R12
-    NOP                                 @ Block #4
-    NOP
-    MLANE   LR, R11, R9, LR
-    NOP
-    NOP
-    NOP
-    NOP
-    BIC     R7, R7, R2, ASR#1
-    STMIA   R5!, {R0, R1, R10, LR}      @ write 4 stereo samples
+    /* Mix the first 4 stereo samples, then the next 4. */
+    .rept 2
+      LDMIA   R5, {R0, R1, R10, LR}       @ load 4 stereo samples to Registers
+      .irp REG, R0, R1, R10, LR           @ 4 blocks
+        MULNE   R9, R7, R12
+        NOP
+        NOP
+        MLANE   \REG, R11, R9, \REG
+        NOP
+        NOP
+        NOP
+        NOP
+        BIC     R7, R7, R2, ASR#1
+      .endr
+      STMIA   R5!, {R0, R1, R10, LR}      @ write 4 stereo samples
+    .endr
 
-    LDMIA   R5, {R0, R1, R10, LR}       @ load the next 4 stereo samples
-    MULNE   R9, R7, R12
-    NOP                                 @ Block #1
-    NOP
-    MLANE   R0, R11, R9, R0
-    NOP
-    NOP
-    NOP
-    NOP
-    BIC     R7, R7, R2, ASR#1
-    MULNE   R9, R7, R12
-    NOP                                 @ Block #2
-    NOP
-    MLANE   R1, R11, R9, R1
-    NOP
-    NOP
-    NOP
-    NOP
-    BIC     R7, R7, R2, ASR#1
-    MULNE   R9, R7, R12
-    NOP                                 @ Block #3
-    NOP
-    MLANE   R10, R11, R9, R10
-    NOP
-    NOP
-    NOP
-    NOP
-    BIC     R7, R7, R2, ASR#1
-    MULNE   R9, R7, R12
-    NOP                                 @ Block #4
-    NOP
-    MLANE   LR, R11, R9, LR
-    NOP
-    NOP
-    NOP
-    NOP
-    BIC     R7, R7, R2, ASR#1
-    STMIA   R5!, {R0, R1, R10, LR}      @ write 4 stereo samples
```

Comment:
I personally don't yet know if I like that simplification (and the ones below). The code is certainly smaller and technically, less prone to errors, and generally more elegant but I'm not sure if it's easier to understand (I personally like to avoid macros in asm since asm is already hard enough to read).